api/v1alpha1/bmcjob_types.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 Tinkerbell.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// BMCJobConditionType represents the condition of the BMC Job.
+type BMCJobConditionType string
+
+const (
+	// JobCompleted represents successful completion of the BMC Job tasks.
+	JobCompleted BMCJobConditionType = "Completed"
+	// JobFailed represents failure in BMC job execution.
+	JobFailed BMCJobConditionType = "Failed"
+	// JobTaskCompleted represents successful completion of a BMC Job task.
+	JobTaskCompleted BMCJobConditionType = "TaskCompleted"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// BMCJobSpec defines the desired state of BMCJob
+type BMCJobSpec struct {
+	// Tasks represents a list of baseboard management actions to be executed.
+	// The tasks are completed in order.
+	Tasks []Task `json:"tasks"`
+}
+
+// Task represents the action to be performed.
+// A single task can only perform one type of action.
+// For example either PowerAction or BootAction.
+// +kubebuilder:validation:MaxProperties:=1
+type Task struct {
+	// PowerAction represents a baseboard management power operation.
+	PowerAction *PowerAction `json:"powerAction,omitempty"`
+
+	// BootAction represents a baseboard management boot device operation.
+	BootAction *BootAction `json:"bootAction,omitempty"`
+}
+
+type PowerAction struct {
+	// State represents the requested power state to set for the baseboard management.
+	// +kubebuilder:validation:Enum=On;Off;Status;Cycle;Reset;Soft
+	State PowerState `json:"state"`
+}
+
+type BootAction struct {
+	// BootDevice represents the requested boot device to set.
+	// +kubebuilder:validation:Enum=PXE;Disk;BIOS;CDROM;Safe
+	BootDevice BootDevice `json:"bootDevice"`
```

Comment:
Can this be modeled as a slice so we can set an order?

Reply:
Should this be just `Device`?

Reply:
Well that depends on if we want to represent it as a slice. I was thinking of `BootOrder`? And for now the controller would just set the first device on the `BootOrder`?
Unless we feel we don't want to do that now and limit to just one Device (first boot device). 

With that said, we could always add to the API in the future and give an option to set single device or entire boot order. Or a new action itself `BootOrderAction`.

Reply:
Given this is wrapped up in a job I think we should model them as 2 separate actions actually. `PersistentBootOrder` and `OneTimeBootDevice`. I suggest that because I think, given the discussion on the persistent flag, we're actually trying to model 2 distinct actions. 

@micahhausler interested in your thoughts on the above too.

Reply:
That sounds good to me 👍  By separating them we dont have to bake in logic later. And we can implement what we want now.